### DISCUSSION TITLE
how to debug the endpoints created by fastapi_users.get_users_router

### DISCUSSION THREAD

**Discussion by lpdswing** [fastapi-users/fastapi-users#1271] [Q&A] [answered]
- I want to get the user id in the middleware so that I can add the user id to the global log. How can I do it? The current user cannot be used in the middleware because it is a dependency.

```python
class GlobalUserMiddleware:
    def __init__(self, app: ASGIApp) -> None:
        self.app = app

    async def __call__(self, scope: Scope, receive: Receive, send: Send) -> None:
        from src.lib.logger import log

        if scope["type"] != "http":
            await self.app(scope, receive, send)
            return
        if scope["type"] == "http":
            request = Request(scope)
            try:
                subject = current_user().id
            except Exception as e:
                log.error(e)
                subject = None
            global_userid.set(subject)
        await self.app(scope, receive, send)
```

- How should I break the built-in endpoints of fastapi_users for debugging

**Comment by frankie567** [comment]
The `current_user` dependency is not designed to work in a middleware. I don't know what you want to do exactly but you can achieve something similar with a global dependency:

```py
async def set_global_user(request: Request, user = Depends(current_user)):
    # request.state is really useful to put shared state in the request object
    # Ref: https://www.starlette.io/requests/#other-state
    request.state.global_user_id = user.id

app = FastAPI(dependencies=[Depends(set_global_user)])
```

**Reply by lpdswing** [reply]
It is not a good way for me to use curren_user directly. I don't want to throw 401 directly when the user is not logged in. Maybe I should parse the auth token directly, but like current_user, get auth router ... is a function returned by the factory , it is difficult to debug through debug mode. thank you for your reply

**Reply by frankie567** [reply, marked as answer]
You can use the `optional` flag in this case:

```py
current_optional_user = fastapi_users.current_user(optional=True)

async def set_global_user(request: Request, user = Depends(current_optional_user)):
    # request.state is really useful to put shared state in the request object
    # Ref: https://www.starlette.io/requests/#other-state
    if user is not None:
        request.state.global_user_id = user.id

app = FastAPI(dependencies=[Depends(set_global_user)])

**Reply by lpdswing** [reply]
Yeay, its work, this is better than parse token directly.
<img width="1563" alt="image" src="https://github.com/fastapi-users/fastapi-users/assets/20874036/0086dec4-0b1b-4ab6-8b04-4554a830ed5a">
